### PR TITLE
[WIP] Bid deduplication enhancement

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
Detect duplicated bids by category only (or hb_pb_dur) and dedupe bid with lower price

### CODE REVIEW

File: exchange/exchange.go [hunk 1]
```diff
@@ -581,15 +582,27 @@ func applyCategoryMapping(ctx context.Context, requestExt openrtb_ext.ExtRequest
...
 			} else {
 				categoryDuration = fmt.Sprintf("%s_%ds", pb, newDur)
+				dupeKey = categoryDuration
 			}
 
-			if dupe, ok := dedupe[categoryDuration]; ok {
+			if dupe, ok := dedupe[dupeKey]; ok {
 				// 50% chance for either bid with duplicate categoryDuration values to be kept
-				if rand.Intn(100) < 50 {
+				dupeBidPrice, dupeBidPriceErr := strconv.ParseFloat(dupe.bidPrice, 64)
```
Comment (SyntaxNode) [inline exchange/exchange.go:596]: Nitpick: The variables name seem a bit long. Consider shortening to `dupePrice` and `err`.

File: exchange/exchange.go [hunk 2]
```diff
@@ -581,15 +582,27 @@ func applyCategoryMapping(ctx context.Context, requestExt openrtb_ext.ExtRequest
...
+				if dupeBidPriceErr != nil {
+					dupeBidPrice = 0
+				}
+				currBidPrice, currBidPriceErr := strconv.ParseFloat(pb, 64)
```
Comment (SyntaxNode) [inline exchange/exchange.go:600]: Nitpick: The variables name seem a bit long. Consider shortening to `currentPrice` and `err`.

File: exchange/exchange_test.go [hunk 3]
```diff
@@ -1196,7 +1196,7 @@ func TestCategoryDedupe(t *testing.T) {
 	cats4 := []string{"IAB1-2000"}
 	bid1 := openrtb.Bid{ID: "bid_id1", ImpID: "imp_id1", Price: 10.0000, Cat: cats1, W: 1, H: 1}
 	bid2 := openrtb.Bid{ID: "bid_id2", ImpID: "imp_id2", Price: 15.0000, Cat: cats2, W: 1, H: 1}
-	bid3 := openrtb.Bid{ID: "bid_id3", ImpID: "imp_id3", Price: 10.0000, Cat: cats1, W: 1, H: 1}
+	bid3 := openrtb.Bid{ID: "bid_id3", ImpID: "imp_id3", Price: 20.0000, Cat: cats1, W: 1, H: 1}
```
Comment (SyntaxNode) [inline exchange/exchange_test.go:1199]: It's unclear to me which parts of this are testing the `dupeKey = category` branch vs the `dupeKey = categoryDuration` branch. Consider separating into separate test cases for clarity.

File: exchange/exchange.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply exchange/exchange.go]: Fixed!

File: exchange/exchange.go (hunk 2 above)
Comment (VeronikaSolovei9) [reply exchange/exchange.go]: Renamed!

File: exchange/exchange_test.go (hunk 3 above)
Comment (VeronikaSolovei9) [reply exchange/exchange_test.go:1172]: Unit tests added

**Comment by VeronikaSolovei9** [comment]
Still needs to work with unit tests. With `TestNoCategoryDedupe` uncommented running into weird issue during CI checks which is not reproducible locally with `./validate.sh --nofmt --cov --race 10`

**Comment by VeronikaSolovei9** [comment]
@camrice Cam, may I ask you to help with CI error? One error gone after rebasing (thank you for this hint!), now it's something else:
`No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.`
 Thank you!

**Comment by camrice** [comment]
It looks like this one picked up some other commits during the rebase. I am working to sync this branch up again. There is at least 1 old commit that shouldn't be included in this PR.

**Comment by VeronikaSolovei9** [comment]
Thank you! Please confirm the priority of this PR, it's probably ok to wait a little and I'll fix when I'm back to work.

**Comment by camrice** [comment]
I did another rebase on this branch and it seems to have cleaned up the commits. Let me know if there's any issues.

File: exchange/exchange_test.go [hunk 4]
```diff
@@ -1388,7 +1392,7 @@ func TestCategoryDedupe(t *testing.T) {
 		bidCategory, adapterBids, rejections, err := applyCategoryMapping(nil, &requestExt, adapterBids, categoriesFetcher, targData)
 
 		assert.Equal(t, nil, err, "Category mapping error should be empty")
-		assert.Equal(t, 2, len(rejections), "There should be 2 bid rejection messages")
+		assert.Equal(t, 3, len(rejections), "There should be 2 bid rejection messages")
```
Comment (hhhjort) [inline exchange/exchange_test.go:1395]: Should the message be "there should be 3 bid rejection messages" ?

**Comment by SyntaxNode** [comment]
To be clear, "deduplication" in this context is selecting a bid among possibly multiple bidders when the targeting keys are the same. It includes a 50/50 random chance to systemically remove bias.

**Pull request merged**